File: Competitive Programming 3/Section 1.3.3/11679 - Sub-prime/Python/main.py
def main():
    while True:
        B,N = map(int,input().split())
        if B == 0 and N == 0 :
            return
        reserve = list(map(int,input().split()))
        for _ in range(0,N):
            a,b,c = map(int,input().split())
            reserve[a-1]-=c
            reserve[b-1]+=c
        ans = 'S'
        for r in reserve :
            if r < 0 :
                ans = 'N'
                break
        print(ans)
main()

File: Competitive Programming 3/Section 1.3.3/11679 - Sub-prime/C++/main.cpp
#include<bits/stdc++.h>

using namespace std;

int reserve[22];
int B,N,a,b,c;

int main(){
    ios_base::sync_with_stdio(0);cin.tie(0);
    while(cin >> B >> N , B || N ){
        for(int i = 1 ; i <= B ; i ++ ){
            cin>>reserve[i];
        }
        for(int i = 1 ; i <= N ; i ++){
            cin >> a >> b >> c ; 
            reserve[a]-=c;
            reserve[b]+=c;
        }
        char ans = 'S';
        for(int i = 1 ; i <= B ; i ++){
            if(reserve[i]<0){
                ans='N';
                break;
            }
        }
        cout << ans << "\n";
    }
    return 0 ; 
}